modules/core/src/main/java/org/locationtech/jts/geom/Coordinate.java

```diff
@@ -47,37 +48,55 @@
    */
   public static final double NULL_ORDINATE = Double.NaN;
   
-  /**
-   * Standard ordinate index values
-   */
+  /** Standard ordinate index value for, where X is 0 */
   public static final int X = 0;
+
+  /** Standard ordinate index value for, where Y is 1 */
   public static final int Y = 1;
+  
+  /**
+   * Standard ordinate index value for, where Z is 2.
+   *
+   * <p>This constant assumes XYZM coordinate sequence definition, please check this assumption
+   * using {@link #getDimension()} and {@link #getMeasures()} before use.
+   */
   public static final int Z = 2;
 
   /**
-   *  The x-coordinate.
+   * Standard ordinate index value for, where M is 3.
+   *
+   * <p>This constant assumes XYZM coordinate sequence definition, please check this assumption
+   * using {@link #getDimension()} and {@link #getMeasures()} before use.
+   */
+  public static final int M = 3;
+  
+  /**
+   * The x-ordinate.
    */
   public double x;
```

Comment:
would it make sense to deprecate direct access to x and y as well? It seems potentially confusing to have `coord.x`, `coord.y` and `coord.getZ()` be the established access pattern

Reply:
I would like to do so, but discussion with @jnh5y and @dr-jts has me considering if having explicit classes, with access to the primitives, is required. Martin quite correctly focuses on these being primary a data structures, rather than an object.

With java value types on the horizon we are likely to change to the `struct` idea from C. Really want to have coordinates represent a block of [xy], [xyz], [xym], [xyzm] if continuous memory, so when we make an array of them everything is co-located in memory.

Reply:
So my design hat agrees with you, prefer getX() and getY(), make them final so the JRE can inline them to avoid performance penalty which is a key concern ...

Reply:
Have to keep public access to Coordinate x and y fields for backwards compatibility.  These could be removed in a future non-compatible version (i,e, JTS 2)

Reply:
Well, unless a big refactoring is done (as I see @jodygarnett has already done for the z field).

Reply:
just because refactor is done for JTS codebase, does not mean I want to accidentally break client code.

Reply:
I really am not sure value types is on the java roadmap ...

Reply:
The problem with marking as Deprecated is that that triggers compiler errors, which can be annoying.  Can comment them as deprecated, though.

Reply:
really? I thought it just made compiler warnings?

From the annotation javadocs:

> A program element annotated &#64;Deprecated is one that programmers
> are discouraged from using, typically because it is dangerous,
> or because a better alternative exists.  Compilers warn when a
> deprecated program element is used or overridden in non-deprecated code.

In anycase I have it commended it as deprecated as the javadoc approach allows a message to be provided

Reply:
Well, yes, warnings not errors.  Still, it's nice to allow compiling warning-free.

Reply:
my $0.02 - probably a good idea to comment and also mark them. the compiler warnings are useful to raise awareness that users should be moving away from the methods, whereas most wouldn't notice a comment change and then face an unexpected compile failure whenever they actually go away.

Reply:
LOL emitting warnings so people change their code is entirely the point.  Then they change their code and the warnings go away

Reply:
I have been in situations where compile-time warnings trigger a rejection of the update to the library, and it is not possible to change the client code.  So I'm very wary of deprecating things until at least a few releases containing the alternative functionality.